### PR TITLE
run_parts for groupadd and groupdel

### PR DESCRIPTION
**Pull request by edneville**
run_parts currently exists in useradd and userdel, this commit mirrors the functionality with groupadd and groupdel

### CODE REVIEW

**Review by smoser — COMMENTED** [review]
Let me first say that I think the idea of this patch is good.  removing a group and still having processes owned by that group is clearly not what the user intended.

Serge asked me to review this from a "shell programming" perspective, and so I've done that.  I think that at some point, he will problably stop asking me to do such things.

As I understand it, the overall goal is to kill any/all processes that have GID as their primary gid or in their Supplementary group list.
Therefore, the script should exit non-zero in either of the following cases:
 a. there are processes that exist with the provided supplementary gid.
 b. the script is unable determine if there are such processes.

(Note: The proposed script only checks Supplementary, I'm assuming that supplementary always contains primary?... that is to say that in /proc/pid/status, the Gid will always be listed in Groups.  Is that correct?)

Any processes that trolls through /proc has to be aware that processes can start and end while it is processing.  The script here reads /proc once.  As such, it is absolutely possible that a process could spawn subprocesses after the read and before the process is killed. That would result in a false-positive exit 'a'.

I've also commented inline about some error conditions that would result in
false positive exit 'b'.

File: etc/shadow-maint/groupdel-pre.d/01-kill_group_procs.sh [hunk 1]
```diff
@@ -0,0 +1,24 @@
+#!/bin/sh
+
+PATH="/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
```
[the remark this answers is not in the document]
Comment (hallyn) [reply etc/shadow-maint/groupdel-pre.d/01-kill_group_procs.sh:3]: etc/shadow-maint/userdel-pre.d/01-kill_user_procs.sh  does this too, and it does make one feel safer.

Note the admin doesn't run this by hand.  It gets run as a side effect of running userdel.  Now you'd like to think they wouldn't run that without sanitizing their path, but...

File: etc/shadow-maint/groupdel-pre.d/01-kill_group_procs.sh [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    fi
+
+    if [ ! -f "/proc/${PROC}/status" ]; then
+        continue
+    fi
+
+    grep -E '^Groups:.*\b'"${GROUPID}"'\b.*' "/proc/${PROC}/status" >/dev/null || continue
+
+    kill -9 "${PROC}"
+done
```
[the remark this answers is not in the document]
Comment (hallyn) [reply etc/shadow-maint/groupdel-pre.d/01-kill_group_procs.sh:23]: It feels like doing this in c (or rust) and using pidfd_send_signal() would be much safer.  Though much less convenient.

(The pid cycling issue *sounds* impossible to hit, but I can think of at least two ways that we would, including simply setting the max_pid to 10 through some other sysctl writing exploit and using some cgroup freezer mis-configuration)

File: etc/shadow-maint/groupdel-pre.d/01-kill_group_procs.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (smoser) [reply etc/shadow-maint/groupdel-pre.d/01-kill_group_procs.sh:3]: It does not make me feel any safer to have N copies of PATH sitting around that can all be wrong in one way or another.  If I have set my PATH to include a fixed/better programs in /opt/bin  such as 'grep' or 'awk' this script has now broken that.

File: etc/shadow-maint/groupdel-pre.d/01-kill_group_procs.sh [hunk 3]
```diff
@@ -0,0 +1,24 @@
+#!/bin/sh
+
+PATH="/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
+GROUPID=`awk -F: '$1 == "'"${SUBJECT}"'" { print $3 }' /etc/group`
```
Comment (smoser) [inline etc/shadow-maint/groupdel-pre.d/01-kill_group_procs.sh:4]: What happens if the user is not in /etc/group ? Does that matter?
The script would currently just not kill any processes.

An existing user might not be in /etc/group due to libnss-extrausers or yellow-pages or something like ldap integration.

**Comment by hallyn** [comment]
> Serge asked me to review this from a "shell programming" perspective, and so I've done that. I think that at some point, he will problably stop asking me to do such things.

Hah!  Not likely.  Thank you.

**Comment by edneville** [comment]
Hello,

On 2023-04-19 08:24-0700, Scott Moser wrote:
> Serge asked me to review this from a "shell programming" perspective, 
> and so I've done that.  I think that at some point, he will problably 
> stop asking me to do such things.

Thanks for taking the time to review!

> As I understand it, the overall goal is to kill any/all processes that 
> have GID as their primary gid or in their Supplementary group list. 

Yes, this is the intention. Ideally if "pkill" were guaranteed to be 
available then handing off the work would be simple.

> Therefore, the script should exit non-zero in either of the following 
> cases:
>  a. there are processes that exist with the provided supplementary gid.
>  b. the script is unable determine if there are such processes.

I should probably have started with set -e.

> (Note: The proposed script only checks Supplementary, I'm assuming 
> that supplementary always contains primary?... that is to say that in 
> /proc/pid/status, the Gid will always be listed in Groups.  Is that 
> correct?)

It is what I observed, I'm not aware of any exceptions to this - Serge?

> Or alternatively something like this:
> ```bash
> grep -E "^Groups:.*\b${GROUPID}\b" | while read -r line; do
>     tmp=${line%%/status:*}
>     pid=${tmp#/proc/}
>     echo pid=$pid
> done
> ```
> 
> that still has the problem of not reading the 'grep' exit code.

What if /bin/sh isn't /bin/bash, or what if bash isn't present?

> +    grep -E '^Groups:.*\b'"${GROUPID}"'\b.*' "/proc/${PROC}/status" /dev/null || continue
> 
> this could fail if grep doesn't support -E. if that happens, the 
> script will just complain print bad grep usage to stderr for every 
> process and then return 0.

That's true. What assumptions should we make, I think earlier an 
assumption was made that the shell is capable of regex, can we assume 
that grep handles -E too? I don't know if that's safe to do so, I would 
/hope/ that someone using a recent shadow library would also have a 
recent grep.

> +    kill -9 "${PROC}"

> nothing is checking the return code of this 'kill'.
> You're probably assuming that you are "real root", and thus a 'kill -9 
> PID' is guaranteed to succeed if PID exists. Is that always the case?

It should only be invoked by root, and I think set -e should be present. 
If the kill fails, then I'd want the script to fail so the problem can 
be sorted out in this case and started again.

> there is also a possibility (very small) that a the pid $PROC died 
> after line 20 above and before line 22.  kill would then exit non-zero 
> but that would be ok.  The worst case scenario is if $PROC died, a new 
> process spawned that re-used that PID and you are now going to kill it 
> even though it did not have PROC in its 'Groups' list.

Also, perhaps that the pid could be reused ... Would hope this is a very 
small chance.

**Comment by smoser** [comment]
Hi, thanks for responding.
I realize you weren't expecting such a response for a PR that followed convention.  I'm sure you were surprised by 100 lines of response to a small shell script.

Its a pet peeve of mine that linux is so dependent on shell scripts that can fail in all sorts of ways.  @hallyn may be happy to take this as it is, and even reasonably so.  Perfection is the enemy of good enough.  I'm hope that we're moving in the direction of more reliable technologies.

> Yes, this is the intention. Ideally if "pkill" were guaranteed to be available then handing off the work would be simple.

It would not surprise me at all if pkill some or all of the same problems.  Unfortunately naive code is not limited to shell.

> I should probably have started with set -e.

set -e has all sorts of shortcomings itself. I personally avoid it and handle errors explicitly.
https://mywiki.wooledge.org/BashFAQ/105

> What if /bin/sh isn't /bin/bash, or what if bash isn't present?

I'm pretty sure all the examples I gave were posix shell.  '${var#}' is posix complient (i think).

> That's true. What assumptions should we make, I think earlier an assumption was made that the shell is capable of regex, can we assume that grep handles -E too? I don't know if that's safe to do so, I would /hope/ that someone using a recent shadow library would also have a recent grep.

its less "recent grep" as "source of grep".  There are several grep implementations in current use.  At very least there are busybox's and gnu.  In either of those cases any recent version probably has '-E'.  You are correct then that its probably a safe assumption, but even then:
 * why use -E if you don't need it
 * you should check for grep failure.

> It should only be invoked by root, and I think set -e should be present. If the kill fails, then I'd want the script to fail so the problem can be sorted out in this case and started again.

hopefully starting it again would "fix it" i guess.  I have personally had experience with 'useradd' and 'userdel' where I had to edit /etc/passwd or /etc/group to get a failed command to continue.  It sure would be nice if it were like a database transaction and could be rolled back.

> there is also a possibility (very small) that a the pid $PROC died after line 20 above and before line 22. kill would then exit non-zero but that would be ok. The worst case scenario is if $PROC died, a new process spawned that re-used that PID and you are now going to kill it even though it did not have PROC in its 'Groups' list.

You probably do *not* want to exit failure if 'kill -9' tries to kill a process that has died between your check invocation of kill.  That is actually reasonably likely to happen.  If you kill a processes parent, that may take out children or vice-versa.  You're definietly in racey territory.

And that is not even considering the malicious process that is *trying* to avoid being killed.

> Also, perhaps that the pid could be reused ... Would hope this is a very small chance.

This probably *is* a small chance, i agree.

**Comment by edneville** [comment]
On 2023-04-21 11:35-0700, Scott Moser wrote:
> Hi, thanks for responding.
> I realize you weren't expecting such a response for a PR that followed 
> convention.  I'm sure you were surprised by 100 lines of response to a 
> small shell script.

Always good to get another pair of eyes.

> Its a pet peeve of mine that linux is so dependent on shell scripts 
> that can fail in all sorts of ways.  @hallyn may be happy to take this 
> as it is, and even reasonably so.  Perfection is the enemy of good 
> enough.  I'm hope that we're moving in the direction of more reliable 
> technologies.

It is a balancing act, for sure.

> its less "recent grep" as "source of grep".  There are several grep 
> implementations in current use.  At very least there are busybox's and 
> gnu.  In either of those cases any recent version probably has '-E'.  
> You are correct then that its probably a safe assumption, but even 
> then:
>  * why use -E if you don't need it
>  * you should check for grep failure.

Ok, I've updated the PR, let me know your thoughts please.

Ed

**Review by smoser — CHANGES_REQUESTED** [review]
I think you prefer the use of 'sed' and 'grep', can you elaborate on why?

thanks so much for being patient with me.

**Comment by edneville** [comment]
On 2023-04-24 07:16-0700, Scott Moser wrote:
> @smoser requested changes on this pull request.
> 
> I think you prefer the use of 'sed' and 'grep', can you elaborate on why?

A deep-seated belief that nags me it won't work on very old AIX/SunOS 
implementations. However, the changes you proposed do work in dash, so 
that's good enough for me now. It'll be hard to change my old ways.

However, you prove that your way is indeed faster, and the chances of 
running into such old sh implementations that cross-over with a modern 
shadow-maint are vanishingly small.

Thanks for the other suggestions too, I've put them in, I think this 
should be what you're after now.

Ed

**Review by smoser — COMMENTED** [review]
I think this is about as good as it can reasonably be in shell.
The only 2 things remaining to complain about are
1. explicit PATH set in the script ; that is done other places in shadow, so even though i disagree it is "right" from that perspective.
2. if GROUPID is not found by the parsing of /etc/group, then the script will not kill any processes but exit 0.  I can't think of a reason to really really care about this, so oh well.

I'll approve this as it is.

**Review by smoser — APPROVED** [review]

**Comment by hallyn** [comment]
Thanks @smoser  and @edneville

**Pull request merged**